vexcl/optimisation.hpp

```diff
@@ -0,0 +1,200 @@
+#ifndef OPTIMISATIONS_HPP
+#define OPTIMISATIONS_HPP
+
+
+#include <vexcl/operations.hpp>
+
+namespace vex {
+//_________________________________________________ restrict keyword
+// The struct will be used as an identificator of restricted vector
+// expressions.
+struct restricted_vector_terminal {};
+
+// Introduce the new kind of expression to Boost.Proto.
+typedef vector_expression<
+    typename boost::proto::terminal< restricted_vector_terminal >::type
+> restricted_vector_expression;
+
+// The actual terminal.
+// This struct will hold a reference to a vector that needs to be declared with
+// 'restrict' qualifier.
+template <class T>
+struct restricted_vector : restricted_vector_expression {
+typedef T value_type; // has to be here for vex::deduce() to work with restricted vectors.
+const vector<T> &v;
+    restricted_vector(const vector<T> &v) : v(v) {}
+};
+
+// A convenience function to be used in user code.
+template <class T>
+inline restricted_vector<T> restricted(const vector<T> &v) {
+    return restricted_vector<T>(v);
+}
+
+// Code generation helpers.
+namespace traits {
+
+// Tell proto it is ok to use these terminals in vexcl expressions.
+template <>
+struct is_vector_expr_terminal< restricted_vector_terminal > : std::true_type {};
+
+// Tell proto that it is not required to unpack the terminal with
+// proto::value() function.
+template <>
+struct proto_terminal_is_value< restricted_vector_terminal > : std::true_type {};
+
+// Declare kernel parameter for the restricted vector.
+// This is the only specification that is different from vector<T>.
+template <typename T>
+struct kernel_param_declaration< restricted_vector<T> > {
+static void get(backend::source_generator &src,
+const restricted_vector<T>&,
```

Comment:
Can you add an indentation here and through the rest of the file (consistent with the rest of the VexCL code)?